### PR TITLE
Better command line argument handling  #3628

### LINKED ISSUE #3628: Better command line argument handling  (Boost Program Options)
For historical reason, we parse command line options in our tools (eth, rlp, ethvm, ...) manually. Let's improve this by using Boost.Program_options (http://www.boost.org/doc/libs/1_63_0/doc/html/program_options.html).

Pay attention for classes like `MinerAux` that _injects_ its options when included. Maybe Boost have a solution for that.

We can leave ethminer untouched if possible.

### PR DESCRIPTION
**Pull request by demon1999**
Resolves #3628

### CODE REVIEW

**Comment by gumb0** [comment]
Thanks a lot for your work! I'll try to look at it next week.

You could rebase it on the latest develop and resolve conflicts meanwhile.

**Comment by pirapira** [comment]
@demon1999 are you still around?

**Comment by demon1999** [comment]
Yes, I am. Now I'm reading about rebase.

**Comment by chfast** [comment]
@demon1999 it would be better to split this PR into smaller parts. Can you make one that adds only boost program_options first. Then a PR that converts options for one of the smaller tools like rlp or ethvm.

**Comment by winsvega** [comment]
right. it is better to be done module by module. Id like to test testeth options first.

**Comment by demon1999** [comment]
Ок, I'll do it when I'll have time.

**Comment by chfast** [comment]
Hi again.
Please also see https://github.com/ethereum/cpp-ethereum/issues/4613.
They are proposed changes to ethvm command line interface, so we might want to do them the same time.

File: eth/AccountManager.cpp [hunk 1]
```diff
@@ -23,6 +23,8 @@
 #include <libdevcore/FileSystem.h>
 #include <libdevcore/CommonIO.h>
 #include <libethcore/KeyManager.h>
+#include <boost/program_options.hpp>
```
Comment (gumb0) [inline eth/AccountManager.cpp]: Are these includes needed? Only includes are added to this file and nothing else.

File: eth/main.cpp [hunk 2]
```diff
@@ -406,403 +320,601 @@ int main(int argc, char** argv)
...
+		    ("format", po::value<string>(), "<format> Set format.")
+		    ("script", po::value<string>(), "<script> Add script.")
+            ("mainnet", "Use the main network protocol.")
+            ("ropsten", "Use the Ropsten testnet.")
+		    ("testnet", "Use the Ropsten testnet.")
+		    ("gas-floor", po::value<string>(), "Set gasFloor")
+            ("private", po::value<string>(), "<name> Use a private chain.")
+            ("test", "Testing mode: Disable PoW and provide test rpc interface.")
+            ("config", po::value<string>(), "<file> Configure specialised blockchain using given JSON information.")
+            ("oppose-dao-fork", "Ignore DAO hard fork (default is to participate).\n")
```
Comment (gumb0) [inline eth/main.cpp]: This can be deleted

File: eth/main.cpp [hunk 3]
```diff
@@ -406,403 +320,601 @@ int main(int argc, char** argv)
...
+			("ask", po::value<string>(), "See description under.")
+			("bid", po::value<string>(), "See description under.")
+			("unsafe-transactions", "Allow all transactions to proceed without verification. EXTREMELY UNSAFE.")
+	;
+	po::options_description clientMining("Client mining");
+	clientMining.add_options()
+			("address,a", po::value<string>(), "<addr>  Set the author (mining payout) address to given address (default: auto).")
+			("author", po::value<string>(), "<addr>  Set the author (mining payout) address to given address (default: auto).")
+			("mining,m", po::value<string>(), "<on/off/number>  Enable mining, optionally for a specified number of blocks (default: off).")
+			("force-mining,f", "Mine even when there are no transactions to mine (default: off).")
```
Comment (gumb0) [inline eth/main.cpp]: This can be deleted

File: eth/main.cpp [hunk 4]
```diff
@@ -406,403 +320,601 @@ int main(int argc, char** argv)
 	bool chainConfigIsSet = false;
 	string configJSON;
 	string genesisJSON;
+    po::options_description clientDefaultMode("Client mode (default)");
+    clientDefaultMode.add_options()
+		    ("format", po::value<string>(), "<format> Set format.")
+		    ("script", po::value<string>(), "<script> Add script.")
+            ("mainnet", "Use the main network protocol.")
```
Comment (gumb0) [inline eth/main.cpp]: Some lines are indented with spaces, please change to tabs everywhere

File: eth/main.cpp (hunk 2 above)
Comment (demon1999) [reply eth/main.cpp]: why?

File: eth/main.cpp (hunk 3 above)
Comment (demon1999) [reply eth/main.cpp]: Why?

File: eth/AccountManager.cpp (hunk 1 above)
Comment (demon1999) [reply eth/AccountManager.cpp]: I think, I don't need it.

File: eth/main.cpp (hunk 2 above)
Comment (gumb0) [reply eth/main.cpp]: This and `-f ` options are not supported anymore.
Probably there're more outdated options, I'll take a closer look later

File: eth/main.cpp (hunk 3 above)
Comment (demon1999) [reply eth/main.cpp]: all this 4 options?

File: eth/main.cpp (hunk 2 above)
Comment (demon1999) [reply eth/main.cpp]: Ok

File: eth/main.cpp (hunk 3 above)
Comment (gumb0) [reply eth/main.cpp]: Only `-f` & `--force-mining`
mining itself is still supported

File: CMakeLists.txt [hunk 5]
```diff
@@ -149,3 +149,8 @@ endif()
 
 # TODO: Split out json_spirit, libscrypt and sec256k1.
 add_subdirectory(utils)
+
+add_executable(main rlp/main.cpp)
+
+add_executable(peer test/unittests/libp2p/peer.cpp)
```
Comment (gumb0) [inline CMakeLists.txt]: Why do we need this?

File: ethkey/main.cpp [hunk 6]
```diff
@@ -82,23 +86,48 @@ int main(int argc, char** argv)
 	setDefaultOrCLocale();
 	KeyCLI m(KeyCLI::OperationMode::ListBare);
 	g_logVerbosity = 0;
-
+	po::options_description generalOptions("General Options");
+	generalOptions.add_options()
+		("verbosity,v", po::value<string>(), "<0 - 9>  Set the log verbosity from 0 to 9 (default: 8).")
+		("version,V", "Show the version and exit.")
+		("help,h",  "Show this help message and exit.")
+	;
```
Comment (gumb0) [inline ethkey/main.cpp]: This formatting looks weird in our style, please leave semicolon on the last line of the statement.

File: ethkey/main.cpp [hunk 7]
```diff
@@ -82,23 +86,48 @@ int main(int argc, char** argv)
 	setDefaultOrCLocale();
 	KeyCLI m(KeyCLI::OperationMode::ListBare);
 	g_logVerbosity = 0;
-
+	po::options_description generalOptions("General Options");
+	generalOptions.add_options()
+		("verbosity,v", po::value<string>(), "<0 - 9>  Set the log verbosity from 0 to 9 (default: 8).")
```
Comment (gumb0) [inline ethkey/main.cpp]: Couldn't we make it `po::value<int>` ?

File: ethkey/main.cpp [hunk 8]
```diff
@@ -26,12 +26,16 @@
...
+#include <boost/program_options/options_description.hpp>
 #include "BuildInfo.h"
 #include "KeyAux.h"
 using namespace std;
 using namespace dev;
 using namespace dev::eth;
 
+namespace po = boost::program_options;
+
 void help()
```
Comment (gumb0) [inline ethkey/main.cpp]: Do we still need this function?

File: ethkey/main.cpp [hunk 9]
```diff
@@ -82,23 +86,48 @@ int main(int argc, char** argv)
...
 	{
 		string arg = argv[i];
 		if (m.interpretOption(i, argc, argv)) {}
-		else if ((arg == "-v" || arg == "--verbosity") && i + 1 < argc)
-			g_logVerbosity = atoi(argv[++i]);
-		else if (arg == "-h" || arg == "--help")
-			help();
-		else if (arg == "-V" || arg == "--version")
-			version();
+		else if (((arg == "-v" || arg == "--verbosity") && i + 1 < argc) || arg == "-h" || arg == "--help" || arg == "-V" || arg == "--version") {
```
Comment (gumb0) [inline ethkey/main.cpp]: opening brace should be on a separate line

File: ethkey/main.cpp [hunk 10]
```diff
@@ -82,23 +86,48 @@ int main(int argc, char** argv)
...
+			argv[ac] = argv[i];
```
Comment (gumb0) [inline ethkey/main.cpp]: Could you explain what are you trying to achieve in this part?
I don't think assiging to `argv` is a very good idea

I think you're trying to get rid of all the options that are handled in `KeyCLI::interpretOption()` and leave only `-v`, `-V`, `-h` in `argv`
Why do you need this? Wouldn't `po::store(po::parse_command_line(ac, argv, generalOptions), vm);` below just ignore options not listed in `generalOptions` ?

Comment (gumb0) [reply ethkey/main.cpp]: Also look at this http://www.boost.org/doc/libs/1_65_1/doc/html/program_options/howto.html#idp437834784

**Comment by demon1999** [comment]
https://github.com/ethereum/cpp-ethereum/pull/4635

**Comment by gumb0** [comment]
@demon1999 Are you still planning to update this PR or should we close this one and you will proceed with several smaller PRs?

**Comment by demon1999** [comment]
You should close it. One of small pr's is 4635

**Pull request closed without merging**